JoinSplit.sol sig validation to ZkAsset

thomas-waite opened the pull request:
## Summary
This PR moves ECDSA signature validation from the JoinSplit.sol validator contract, where it is currently performed, into the `ZkAsset.sol` `confidentialTransfer()` method.  

## Motivation
This change is motivated by a desire to create a distinction between the role of the validator contracts and the role of the `ZkAsset.sol` contract. The validator contracts determine the mathematical legitimacy of a transaction, whilst the `ZkAsset.sol` contract determines whether a transaction **should** happen. `ZkAsset.sol` performs the permissioning checks. 

## Where to focus/what I'd like feedback on
This PR includes a breaking change (described below), the `confidentialTransfer()` function now takes a second argument `bytes memory signatures`. 

This is necessary, as I have chosen not to encode the `inputSignatures` into the `proofData` (and therefore can't just supply the `proofData` and extract signatures inside `confidentialTransfer()`). 

I did this for gas optimisation reasons. The JoinSplit.sol validator no longer requires signatures, so sending them in the `proofData` would result in redundant information being sent. It costs **2,176 gas** to add 32 bytes of non-zero data to a transaction as input data. Signatures are 96 bytes long, so say the joinSplit has 2 input notes (many more are possible/likely) and this results in an additional **13,000 gas** cost.

Interested in your thoughts on the resulting breaking change and API change.

<!--- Describe your changes in detail -->
## Detailed changes
_aztec.js_
- Updated `inputCoder.js` and `encoderFactory.js` to remove `inputSignatures` from the ABI proof data encoding
- `inputSignatures` still generated inside the `encodeJoinSplitTransaction()` proof construction method, however adapted it to output the signatures as a third argument - rather than integrate into `proofData`. These signatures can then be passed when needed into the relevant smart contract method

_protocol_
- Removed signature validation from `JoinSplit.sol` and updated call data maps
- Performed EIP712 signature validation in the `confidentialTransfer()` method of `ZkAsset.sol`. This is the unilateral transfer method, for which only joinSplit proofs are possible
- Wrote `extractSignature()` and `validateSignature()` functions in `ZkAsset.sol` 
- Updated `confidentialTransfer()` method of `ZkAssetMintable.sol` to match new interface

<!--- Please describe how reviewers can test your changes -->

## Types of changes
**Breaking change**
The interface for `confidentialTransfer()` in ZkAsset.sol has changed. It was previously:

 `function confidentialTransfer(bytes memory _proofData)`

and is now:
 `function confidentialTransfer(bytes memory _proofData, bytes memory _signatures)`

There is a second input argument, which takes an array of ECDSA signatures over all inputNotes which give permission for those notes to be used in a `confidentialTransfer()` operation. 

## Checklist:

<!--- The following points should be used to indicate the progress of your PR.  Put an `x` in all the boxes that apply right now, and come back over time and check them off as you make progress.  If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

-   [x] Prefix PR title with `[WIP]` if necessary.
-   [x] Add tests to cover changes as needed.
-   [x] Update documentation as needed.

_aztec.js_
-   [x] Remove `inputSignatures` from being in the joinSplit `proofData`, update `encoderFactory.js` and `InputCoder.js` to reflect this change
-   [x] Update `encodeJoinSplitTransaction()` to return `inputSignatures` as a third output - these can then be passed to the relevant smart contract method

_protocol_
-  [x] Remove signature validation from `JoinSplit.sol`, and update call data maps
-  [x] Add `bytes memory _signature` as a second input argument to the `ZkAsset.sol` method `confidentialTransfer()`
-  [x] Write `extractSignature()` function in ZkAsset.sol
-  [x] Write `validateSignature()` function in ZkAsset.sol, to validate an EIP712 signature in the ZkAsset domain 
-  [x] Perform signature validation in `confidentialTransfer()`, before get to `confidentialTransferInternal()`
-  [x] Update `ZkAssetMintable.sol` to use new `confidentialTransfer()` interface
-  [x] Write tests to check behaviour, specifically:
      - [x] JoinSplit.sol will perform validation with a fake signature (i.e. confirm it is no longer validating signatures)
      - [x] `confidentialTransfer()` in ZkAsset.sol will successfully validate a correct signature 
      - [x] `confidentialTransfer()` in ZkAsset.sol will fail for no signature 
      - [x] `confidentialTransfer()` in ZkAsset.sol will fail for a fake signature
      - [x] `confidentialTransfer()` in ZkAsset.sol will fail for an incorrect validator address (resulting in the EIP712 domain hash being incorrect)

dependabot-support committed:
build(deps-dev): bump @0x/sol-trace from 2.0.10 to 2.0.11

Bumps [@0x/sol-trace](https://github.com/0xProject/0x-monorepo) from 2.0.10 to 2.0.11.
- [Release notes](https://github.com/0xProject/0x-monorepo/releases)
- [Commits](https://github.com/0xProject/0x-monorepo/compare/@0x/sol-trace@2.0.10...@0x/sol-trace@2.0.11)

Signed-off-by: dependabot[bot] <support@dependabot.com>

dependabot-support committed:
build(deps-dev): bump lerna from 3.13.1 to 3.13.4

Bumps [lerna](https://github.com/lerna/lerna/tree/HEAD/core/lerna) from 3.13.1 to 3.13.4.
- [Release notes](https://github.com/lerna/lerna/releases)
- [Changelog](https://github.com/lerna/lerna/blob/master/core/lerna/CHANGELOG.md)
- [Commits](https://github.com/lerna/lerna/commits/v3.13.4/core/lerna)

Signed-off-by: dependabot[bot] <support@dependabot.com>

dependabot-support committed:
build(deps-dev): bump webpack-cli from 3.3.0 to 3.3.1

Bumps [webpack-cli](https://github.com/webpack/webpack-cli) from 3.3.0 to 3.3.1.
- [Release notes](https://github.com/webpack/webpack-cli/releases)
- [Changelog](https://github.com/webpack/webpack-cli/blob/master/CHANGELOG.md)
- [Commits](https://github.com/webpack/webpack-cli/compare/v.3.3.0...v.3.3.1)

Signed-off-by: dependabot[bot] <support@dependabot.com>

PaulRBerg committed:
Merge pull request #100 from AztecProtocol/dependabot/npm_and_yarn/lerna-3.13.4

build(deps-dev): bump lerna from 3.13.1 to 3.13.4

PaulRBerg committed:
Merge pull request #102 from AztecProtocol/dependabot/npm_and_yarn/webpack-cli-3.3.1

build(deps-dev): bump webpack-cli from 3.3.0 to 3.3.1

PaulRBerg committed:
Merge pull request #99 from AztecProtocol/dependabot/npm_and_yarn/@0x/sol-trace-2.0.11

build(deps-dev): bump @0x/sol-trace from 2.0.10 to 2.0.11

dependabot-support committed:
build(deps-dev): bump web3 from 1.0.0-beta.50 to 1.0.0-beta.52

Bumps [web3](https://github.com/ethereum/web3.js) from 1.0.0-beta.50 to 1.0.0-beta.52.
- [Release notes](https://github.com/ethereum/web3.js/releases)
- [Commits](https://github.com/ethereum/web3.js/compare/v1.0.0-beta.50...v1.0.0-beta.52)

Signed-off-by: dependabot[bot] <support@dependabot.com>

PaulRBerg committed:
Merge pull request #97 from AztecProtocol/dependabot/npm_and_yarn/web3-1.0.0-beta.52

build(deps-dev): bump web3 from 1.0.0-beta.50 to 1.0.0-beta.52

dependabot-support committed:
build(deps-dev): bump truffle from 5.0.10 to 5.0.14

Bumps [truffle](https://github.com/trufflesuite/truffle) from 5.0.10 to 5.0.14.
- [Release notes](https://github.com/trufflesuite/truffle/releases)
- [Commits](https://github.com/trufflesuite/truffle/compare/truffle@5.0.10...truffle@5.0.14)

Signed-off-by: dependabot[bot] <support@dependabot.com>

dependabot-support committed:
build(deps): bump web3-utils from 1.0.0-beta.50 to 1.0.0-beta.52

Bumps [web3-utils](https://github.com/ethereum/web3.js) from 1.0.0-beta.50 to 1.0.0-beta.52.
- [Release notes](https://github.com/ethereum/web3.js/releases)
- [Commits](https://github.com/ethereum/web3.js/compare/v1.0.0-beta.50...v1.0.0-beta.52)

Signed-off-by: dependabot[bot] <support@dependabot.com>

dependabot-support committed:
build(deps-dev): bump truffle-hdwallet-provider from 1.0.6 to 1.0.8

Bumps [truffle-hdwallet-provider](https://github.com/trufflesuite/truffle) from 1.0.6 to 1.0.8.
- [Release notes](https://github.com/trufflesuite/truffle/releases)
- [Commits](https://github.com/trufflesuite/truffle/compare/truffle-hdwallet-provider@1.0.6...truffle-hdwallet-provider@1.0.8)

Signed-off-by: dependabot[bot] <support@dependabot.com>

PaulRBerg committed:
Merge pull request #101 from AztecProtocol/dependabot/npm_and_yarn/web3-utils-1.0.0-beta.52

build(deps): bump web3-utils from 1.0.0-beta.50 to 1.0.0-beta.52

PaulRBerg committed:
Merge pull request #103 from AztecProtocol/dependabot/npm_and_yarn/truffle-hdwallet-provider-1.0.8

build(deps-dev): bump truffle-hdwallet-provider from 1.0.6 to 1.0.8

PaulRBerg committed:
Merge pull request #96 from AztecProtocol/dependabot/npm_and_yarn/truffle-5.0.14

build(deps-dev): bump truffle from 5.0.10 to 5.0.14

thomas-waite committed:
feat(protocol): remove signature validation from JoinSplitABIEncoder.sol

thomas-waite committed:
test(protocol): remove JoinSplitABIEncoder.sol signature logic tests

thomas-waite committed:
test(protocol): metacommit demonstrating required flow

thomas-waite committed:
Merge branch 'develop' of github.com:AztecProtocol/AZTEC into develop

thomas-waite committed:
Merge branch 'develop' of github.com:AztecProtocol/AZTEC into develop

thomas-waite committed:
feat(protocol): remove inputSignatures from calldata map/input

thomas-waite committed:
refactor(aztec.js): remove input signatures from JoinSplit inputCoder

thomas-waite committed:
refactor(aztec.js): remove signature construction from proof construct

thomas-waite committed:
test(protocol): add joinSplit test for wrong note owners

thomas-waite committed:
refactor(aztec.js): re-add signature generation to proof construction

Done as the encodeJoinSplitTransaction is a conveninent place to generate the signatures
Means that the user only has to call this function
Signatures then returned as a third output argument

thomas-waite committed:
test(protocol): update tests to use new joinSplit inputCoder API

thomas-waite committed:
feat(protocol): introduce a validateSignature() function

Split out the confidentialApprove() function into two - confidentialApprove() and validateSignature()
Done as validating signature functionality will now be needed in multiple places

thomas-waite committed:
refactor(aztec.js): concatenate all sigs in joinSplit proof construction

thomas-waite committed:
refactor(protocol): perform signature validation in ZkAsset

the confidentialTransferInternal() method now takes two arguments: proofOutputs and _signatures.
This allows signature validation to be performed in the ZkAsset rather than the JoinSplit.sol
verifier

BREAKING CHANGE: the confidentialTransfer() method in ZkAsset.sol now requires 2 input arguments
(proofOutputs, _signatures) rather than just 1 (proofOutputs)

thomas-waite committed:
test(protocol): add ZkAsset permissioning test for new sig validation

thomas-waite committed:
test(protocol): update tests with new confidentialTransfer() interface

thomas-waite committed:
refactor(aztec.js): format signature components in proof construction

Slice all '0x' off the (v, r, s) components
Make sure to return an empty array, rather than nothing/undefined, in the case where signatures are not generated

thomas-waite committed:
refactor(protocol): update interface to match confidentialTransfer()

thomas-waite committed:
feat(protocol): write validateSignature() function in ZkAsset.sol

function validates an EIP712 signature for a JoinSplit typeHash
write a helper function extractSignature() to extract the signature for a particular input note from an array of inputSignatures
calculate the JOIN_SPLIT_SIGNATURE_TYPE_HASH

thomas-waite committed:
style(protocol): tidy up - remove .only, chmod trace.sh

thomas-waite committed:
refactor(aztec.js): generate domain params for ZkAsset rather than ACE

thomas-waite committed:
test(protocol): failure tests for fake and no signatures, tidy up

thomas-waite committed:
refactor(protocol): calculate challenge in confidentialTransferInternal

use challenge to construct the structHash part of the EIP712 signature
pass proofData into confidentialTransferInternal, to get access to the challenge
general tidying up and removing of unneeded events

thomas-waite committed:
test(protocol): update tests for new confidentialTransfer() interface

thomas-waite committed:
refactor(protocol): change validateSignature() modifier to internal

thomas-waite committed:
refactor(protocol): make confidentialTransfer call interface consistent

thomas-waite committed:
test(aztec.js): update abiEncoder tests with proof construction refactor

thomas-waite committed:
style(protocol): eslint and prettier fixes

thomas-waite committed:
test: update validatorAddress in test proof construction

thomas-waite committed:
Merge branch 'develop' of github.com:AztecProtocol/AZTEC into develop

thomas-waite committed:
refactor: fix naming and import issues after rebase

thomas-waite committed:
refactor(protocol): rename sig flow test file and update test

thomas-waite committed:
refactor(aztec.js): enable empty validatorAddress and inputNoteOwners

thomas-waite committed:
test(protocol): tests for none and fake signatures

thomas-waite committed:
test(protocol): tests for no validatorAddress/inputNoteOwners

thomas-waite committed:
style(aztec.js): update docs

thomas-waite committed:
style(protocol): update docs for new functions

thomas-waite committed:
refactor(protocol): remove all LIBEIP712 inheritance and sig reference

thomas-waite committed:
refactor(protocol): remove all LIEIP712 inheritance and sig calculation

thomas-waite committed:
refactor(protocol): use validateSig function in confidentialApprove

ArnSch commented on packages/aztec.js/src/abiEncoder/inputCoder.js:
```diff
@@ -32,6 +32,18 @@ inputCoder.bilateralSwap = (proofData, challenge, owners, metadata) => {
+/**
+ * Input encoding for burn proof data
+ *
+ * @method mint
```
```suggestion
 * @method burn
```

ArnSch replied:
(also, these docs are not very informative, but presume this is going to change a lot anyways through Paul's work)

thomas-waite replied:
Done, and yep I think so

ArnSch commented on packages/aztec.js/src/proof/joinSplit/index.js:
```diff
@@ -219,33 +225,29 @@ joinSplit.encodeJoinSplitTransaction = ({
+        });
+        signatures = `0x${signaturesArray.join('')}`;
+    } else {
+        signatures = [];
```
I think we want proof construction to exit if we have `inputNotes` but no `validatorAddress` or no `inputNoteOwners`, there is no point in proceeding and having the user have a hard to digest error about the proof not being valid if we can give more helpful feedback on client side.

This also simplifies the logic, since you can check if an error needs to be thrown, and if not just go through one code path (the map).

thomas-waite replied:
As discussed. 

Implementing a check for if `inputNoteOwners.length > 0 && validatorAddress.length === 0`, then throw. 

Otherwise, proceed. 

Replaced `inputNotes.map()` with `inputNoteOwners.map()`

ArnSch commented on packages/protocol/contracts/ERC1724/ZkAsset.sol:
```diff
@@ -131,49 +204,56 @@ contract ZkAsset is IZkAsset, IAZTEC, LibEIP712 {
-        if (publicValue > 0) {
-            emit RedeemTokens(publicOwner, uint256(publicValue));
+    /**
+    * @dev Perform ECDSA signature validation for a signature over an input note
```
Minor: not a great description of confidentialTransferInternal

thomas-waite replied:
Good point, done

ArnSch commented on packages/protocol/test/ACE/validators/joinSplit/signatureFlow.js:
```diff
@@ -0,0 +1,195 @@
+/* global artifacts, expect, contract, beforeEach, it:true */
```
Can you double check that these tests abide by @PaulRBerg's new standards?

ArnSch replied:
Specifically, I know Paul recently standardised how tests are named.

thomas-waite replied:
Chatted with him, you're right. Updated test naming

ArnSch reviewed (changes requested):
A couple things which are worth changing, but otherwise 👍

thomas-waite committed:
style(protocol): update test naming to use new standard

thomas-waite committed:
style: doc updates and unneeded comment removal

thomas-waite committed:
test(protocol): move sig permissioning tests to ZkAsset.js

delete ZkAssetPermissioning.js and move the tests to ZkAsset.js
done to follow test standard

thomas-waite committed:
test(protocol): move signatureFlow for joinSplit validator into index.js

done to be consistent with test standards

thomas-waite committed:
style(aztec.js): docs for proofUtils.randomAddress()

thomas-waite committed:
refactor(aztec.js): implement PR feedback

throw error if inputNoteOwners.length > 0 and validatorAddress.length === 0
perform .map for signatureArray over inputNoteOwners, rather than inputNotes

thomas-waite committed:
test(protocol): update tests to only input required noteOwners

thomas-waite committed:
refactor(dev-utils): new errorType code

thomas-waite committed:
style: eslint and prettier fixes

thomas-waite committed:
refactor(aztec.js): move signature param checks to proofUtils

also check that if inputNoteOwners is non-zero, there are inputNotes also passed in

thomas-waite committed:
test(aztec.js): test error thrown for incorrect signature params

thomas-waite committed:
style: eslint and prettier fixes

thomas-waite committed:
test(protocol): unravel beforeEach(), move content into unit tests

thomas-waite committed:
test(protocol): implement proper confidentialApprove() flow

remove inputNoteOwners accounts from proof construction - to be sure that confidentialApprove() is being used to grant permission, and that signatures are not being seperately validated
done to thoroughly test confidentialApprove()

thomas-waite committed:
test(aztec.js): add failure test for if confidentialApprove() not called

thomas-waite committed:
style: apply prettier fixes
